src/facebook_client.py

```diff
@@ -0,0 +1,128 @@
+        if params is None:
+            params = {}
+        params = params.copy()
+        params["access_token"] = self._access_token
+
+        url = f"{_BASE_URL}{endpoint}"
+        response = requests.get(url, params)
+
+        if "data" not in response.json():
+            log.error(f"Response from Facebook did not contain a 'data' field. "
+                      f"The returned data is probably an error message: {response.json()}")
+            exit(1)
+
+        return response
+
+    def _make_paged_get_request(self, endpoint, params=None):
+        response = self._make_get_request(endpoint, params)
+        result = response.json()["data"]
+
+        next_url = response.json().get("paging", {}).get("next")
+        while next_url is not None:
+            response = requests.get(next_url)
+            result.extend(response.json()["data"])
+            next_url = response.json()["paging"].get("next")
+        return result
+
+    def get_all_posts_published_by_page(self, page_id, fields=["attachments", "created_time", "message"]):
+        log.debug(f"Fetching all posts published by page '{page_id}'...")
+        posts = self._make_paged_get_request(
+            f"/{page_id}/published_posts",
+            {
+                "fields": ",".join(fields),
+                "limit": _MAX_LIMIT
+            }
+        )
+        log.info(f"Fetched {len(posts)} posts")
+        return posts
+
+    def get_top_level_comments_on_post(self, post_id, fields=["attachments", "created_time", "message"]):
+        return self._make_paged_get_request(
+            f"/{post_id}/comments",
+            {
+                "fields": ",".join(fields),
+                "limit": _MAX_LIMIT,
+                "filter": "toplevel"
+            }
+        )
+
+    def get_all_comments_on_post(self, post_id, fields=["parent", "attachments", "created_time", "message"]):
+        log.info(f"Fetching all comments on post '{post_id}'...")
+        comments = self._make_paged_get_request(
+            f"/{post_id}/comments",
+            {
+                "fields": ",".join(fields),
+                "limit": _MAX_LIMIT,
+                "filter": "stream"
+            }
+        )
+        log.info(f"Fetched {len(comments)} comments")
+        return comments
+
+    def get_all_comments_on_page(self, page_id, fields=["parent", "attachments", "created_time", "message"]):
+        log.info(f"Fetching all comments on page '{page_id}'...")
+        posts = self.get_all_posts_published_by_page(page_id, fields=["id", "is_inline_created"])
+
+        # Posts that were used as adverts are returned twice by Facebook, one post representing the page post and
+        # one representing the advert. Both 'posts' have a comments edge pointing to the same dataset, so reading
+        # both would lead to duplicated comments in the returned data.
+        # We solve this by pre-filtering for posts that were not created inline i.e. the original page post only.
```

Comment:
```suggestion
        # We solve this by pre-filtering for posts that were not created inline i.e. return the original page post only.
```

Reply:
Changed to 'i.e. we filter for the original page posts only.' (because we're not strictly returning anything here, but appreciate that the original wording was a bit unclear)